zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+        type MaxOwners: Get<u32>;
+
+        /// The minimum required amount to vote on an outcome.
+        #[pallet::constant]
+        type MinOutcomeVoteAmount: Get<BalanceOf<Self>>;
+
+        /// The maximum number of keys to remove from a storage map.
+        #[pallet::constant]
+        type RemoveKeysLimit: Get<u32>;
+
+        /// The vote lock identifier.
+        #[pallet::constant]
+        type VoteLockIdentifier: Get<LockIdentifier>;
+
+        /// The fee required to add a voting outcome.
+        #[pallet::constant]
+        type VotingOutcomeFee: Get<BalanceOf<Self>>;
+
+        type WeightInfo: WeightInfoZeitgeist;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Sender tried to vote with an amount below a defined minium.
+        AmountTooLow,
+        /// The global dispute period is already over and the winner is determined.
+        GlobalDisputeAlreadyFinished,
+        /// Sender does not have enough funds for the vote on an outcome.
+        InsufficientAmount,
+        /// No global dispute present at the moment.
+        NoGlobalDisputeStarted,
+        /// The voting outcome has been already added.
+        OutcomeAlreadyExists,
+        /// The outcome specified is not present in the voting outcomes.
+        OutcomeDoesNotExist,
+        /// The global dispute period is not over yet. The winner is not yet determined.
+        UnfinishedGlobalDispute,
+        /// The maximum number of votes for this account is reached.
+        MaxVotesReached,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(fn deposit_event)]
+    pub enum Event<T>
+    where
+        T: Config,
+    {
+        /// A new voting outcome has been added. \[market_id, outcome_report\]
+        AddedVotingOutcome(MarketIdOf<T>, OutcomeReport),
+        /// The winner of the global dispute system is determined. \[market_id\]
+        GlobalDisputeWinnerDetermined(MarketIdOf<T>),
+        /// The outcome owner has been rewarded. \[market_id\]
+        OutcomeOwnerRewarded(MarketIdOf<T>),
+        /// The outcomes storage item is partially cleaned. \[market_id\]
+        OutcomesPartiallyCleaned(MarketIdOf<T>),
+        /// The outcomes storage item is fully cleaned. \[market_id\]
+        OutcomesFullyCleaned(MarketIdOf<T>),
+        /// A vote happened on an outcome. \[market_id, outcome, vote_amount\]
+        VotedOnOutcome(MarketIdOf<T>, OutcomeReport, BalanceOf<T>),
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn reward_account(market_id: &MarketIdOf<T>) -> T::AccountId {
+            T::GlobalDisputesPalletId::get().into_sub_account(market_id)
+        }
+    }
+
+    impl<T> GlobalDisputesPalletApi<MarketIdOf<T>, AccountIdOf<T>, BalanceOf<T>> for Pallet<T>
+    where
+        T: Config,
+    {
+        fn push_voting_outcome(
+            market_id: &MarketIdOf<T>,
+            outcome: OutcomeReport,
+            owner: &T::AccountId,
+            vote_balance: BalanceOf<T>,
```

Comment:
That is a great point! I developed a pallet specifically for the current prediction market pallet logic. I saw the problem, that disputors could be really disappointed, that there hard earned and big `DisputeBond reserves` doesn't have an effect on the global dispute vote, because they are `reserved` and can't be used for another lock (`reserves` can't overlap with `locks`). 
In addition to that, I wanted to determine the voting winner, even if no vote happened with this `vote_balance`. Like in example for simple_disputes the global dispute mechanism would initially resolve to the last dispute outcome, because it had the highest dispute bond. 

But yes, the more I think about this, the more I think of giving it an `Option` and name it `initial_vote_balance`. Since we control the private API, we can have this flexibility. But yeah, I should comment that too, because I assume inside the `zrml-global-dispute` that there exists an already "paid" `initial_vote_balance` (in some kind or form of a reserve, transfer or some other kind of value transfer).

Reply:
I think it would make sense to open the possibility to pass the `vote_balance` and let the global disputes pallet handle it. The current logic rewards the correct outcome owners with the funds from the incorrect outcome owners. Having that option would allow use to just push all `DisputeBonds` to the global dispute pallet, which will eventually properly distribute them to the rightful owners.

Reply:
Do you requested changes with that feedback? I think it's exactly as implemented.

Just a quick addition: For the private API we allow multiple owners. For the public dispatch `add_vote_outcome` we allow only one owner to reduce the overall storage, since multiple owners are expensive to store. In addition to that, it would make not much sense to allow multiple outcome owners for the public dispatch, because it means, that somebody can spam the same correct outcome to get more share of the loosing fees.

Apart from that, I realised that passing an `initial_vote_balance: Option<BalanceOf<T>>` wouldn't make much sense, because the caller can just specify `Zero::zero()` as `initial_vote_balance` instead of `None`.

1b4a8309c29f6a21b025224c7acc94cb5d743c63